File: 0049-group-anagrams/0049-group-anagrams.java
class Solution {
    public List<List<String>> groupAnagrams(String[] strs) {
		Map<String, List<String>> result = new HashMap<>();

		for (String s : strs) {
			char[] charArray = s.toCharArray();
			// 단어 사전순 정렬
			Arrays.sort(charArray);

			String key = String.valueOf(charArray);
			// list 초기화
			if (!result.containsKey(key)) {
				result.put(key, new ArrayList<>());
			}
			result.get(key).add(s);
		}

		return new ArrayList<>(result.values());
    }
}

File: 0049-group-anagrams/0049-group-anagrams.kt
class Solution {
    fun groupAnagrams(strs: Array<String>): List<List<String>> {
        // 애너그램 보관 map
        val result: MutableMap<String, MutableList<String>> = mutableMapOf()

        for (s in strs) {
            // 문자열 정렬
            val key = s.toCharArray().sorted().joinToString()

            // map key 등록
            result.getOrPut(key) { mutableListOf() }

            // null 아님
            result[key]!!.add(s)
        }

        return ArrayList<List<String>>(result.values)
    }
}